### PR TITLE
Labels infrastructure overhaul

### PR DESCRIPTION
**Pull request by brimoor**
**Release Notes**
- introducing a base `eta.core.labels.Labels` class from which all labels derive
- introducing a base `eta.core.labels.LabelsSchema` from which all labels schemas derive
- introducing a base `eta.core.labels.LabelsFrameRenderer` class that encapsulates the action of rendering information about spatiotemporal entities for an individual frame
- adding a base `eta.core.frames.FrameLabels` class to encapsulate `Labels` classes that describe content in individual images/video frames
- adding a new `eta.core.objects.Object` class that represents a spatiotemporal object in a video and all of its per-frame observations
- adding support for frame-level labels in `eta.core.events.Event`s 
- adding a new `eta.core.events.DetectedEvent` class to encapsulate per-frame information about an event
- adding support for `eta.core.events.Event`s to contain `Object`s and other `Event`s
- adding support for rendering `Object`s, `Event`s, and `VideoLabels` at the frame-level via the `LabelsFrameRenderer` interface

**Important Comments**
- EVERYTHING related to `Labels` in this PR is backwards compatible. All existing `VideoLabels` are still valid `VideoLabels`, for example
- The format of schemas has changed slightly. Previously serialized schemas will no longer be readable by their new `LabelsSchema` implementations

**TODO**
- add support for rendering ImageLabels/VideoLabels via `eta.core.annotations`
- add support for converting framewise Object/Event/VideoLabels to spatiotemporal format
- finish implementing schema validation for objects/events with children

**Test Code**

```py
from copy import deepcopy

import eta.core.data as etad
import eta.core.image as etai
import eta.core.objects as etao
import eta.core.video as etav

import theta.core.objects as theo

'''
# Download some test data to work with
eta gdrive download 129Vw0kkVY9eOBscdMcBLswx1ov6rXae6 /path/to/animals.json
eta gdrive download 1T82D1RUfRqpXKtzYwxBeEGyBXUIwpqLU /path/to/vehicles.json
'''

video_labels_path = "/path/to/vehicles.json"
image_set_labels_path = "/path/to/animals.json"

video_labels = etav.VideoLabels.from_json(video_labels_path)

schema = video_labels.get_active_schema()
print(schema)

#
# Convert DetectedObjects to Objects with object-level attributes
#

video_labels2 = etav.VideoLabels()

for traj in theo.TrajectorySet.from_video_labels(video_labels):
    obj = etao.VideoObject()

    all_attrs = etad.AttributeContainer()
    for dobj in traj.iter_sorted_objects():
        obj.label = dobj.label
        obj.index = dobj.index

        all_attrs.add_container(dobj.attrs)
        dobj_clean = deepcopy(dobj)
        dobj_clean.clear_attributes()
        obj.add_detection(dobj_clean, frame_number=dobj_clean.frame_number)

    static_attrs = etad.majority_vote_categorical_attrs(all_attrs)
    obj.add_object_attributes(static_attrs)
    video_labels2.add_object(obj)

schema2 = video_labels2.get_active_schema()
print(schema2)

#
# Test exclusive attributes
#

# Verify that VideoLabels are compliant with their own active schema
schema2.validate(video_labels2)

# Make `make` an exclusive attribute, and add a duplicate `make` to one frame
schema3 = deepcopy(schema2)
schema3.objects.schema["vehicle"].attrs.schema["make"].exclusive = True

video_labels3 = deepcopy(video_labels2)

done = False
for oidx, obj in enumerate(video_labels3.iter_video_objects()):
    for aidx, attr in enumerate(obj.attrs):
        if attr.name == "make" and not done:
            obj.attrs.add(deepcopy(attr))
            done = True
            print(
                "Added duplicate attribute (oidx, aidx) = (%d, %d)" %
                (oidx, aidx))

# Verify that something happened
print("video_labels2: %d" % len(video_labels2.to_str()))
print("video_labels3: %d" % len(video_labels3.to_str()))

schema3.validate(video_labels2)  # No error
#schema3.validate(video_labels3)  # Error

# Remove the duplicate `make` attribute
video_labels3.filter_by_schema(schema3)

# Verify that extra attribute was removed
print("video_labels2: %d" % len(video_labels2.to_str()))
print("video_labels3: %d" % len(video_labels3.to_str()))

# Test schema subsets
schema4 = deepcopy(schema)
schema4.merge_schema(schema2)

schema.validate_subset_of_schema(schema4)  # No error
schema2.validate_subset_of_schema(schema4)  # No error
schema4.validate_subset_of_schema(schema4)  # No error

#schema4.validate_subset_of_schema(schema)  # Error
#schema4.validate_subset_of_schema(schema2)  # Error

# Test framewise rendering of Objects
video_labels4 = video_labels2.render_framewise_labels()
print("video_labels2:\n%s\n" % video_labels2)
print("video_labels4:\n%s\n" % video_labels4)

#
# Convert some objects to members of events
#

video_labels5 = etav.VideoLabels()

event = etae.VideoEvent(index=1)
for obj in video_labels2.objects:
    event.add_object(obj)

video_labels5.add_event(event)

# Test framewise rendering of Events
video_labels6 = video_labels5.render_framewise_labels()
print("video_labels5:\n%s\n" % video_labels5)
print("video_labels6:\n%s\n" % video_labels6)
```

### CODE REVIEW

**Comment by brimoor** [comment]
Okay I'll add audit support in a separate PR.

I don't want to remove `filter_by_schema()` yet. It's valid functionality. We can remove it in whatever PR migrates that functionality to a new home

**Comment by tylerganter** [comment]
I would prefer for `exclusive` to be serialized even when `false`. It's a bit easier to work with visually

Also wondering why `BooleanAttributeSchema.values` is a thing. Can't we just assume that it's `true/false`?

**Comment by brimoor** [comment]
`BooleanAttributeSchema.values` allows you to know whether `True` or `False` are actually observed in your labels. Which might now happen if all instances are either `True` or `False`

**Comment by brimoor** [comment]
The codebase also follows a convention that falsey things are omitted from serialization whenever possible. This enables, for example, serialized JSON from a newer version of ETA to be ingested by an older version of ETA that isn't aware of fancy new features. I felt that omitting `exclusive` whenever possible matched that spirit

**Comment by brimoor** [comment]
FYI I added example code to the PR description that I've been using to test the various functionality

**Comment by brimoor** [comment]
Alright `exclusive` is always serialized now

File: eta/core/objects.py [hunk 1]
```diff
@@ -233,20 +338,412 @@ def sort_by_frame_number(self, reverse=False):
...
+        '''Whether the object has at least one child Object.'''
+        return bool(self.child_objects)
+
+    def iter_attributes(self):
+        '''Returns an iterator over the object-level attributes of the object.
+
+        Returns:
+            an iterator over `Attribute`s
+        '''
+        return iter(self.events)
```
Comment (tylerganter) [inline eta/core/objects.py]: `self.attrs`

Comment (brimoor) [reply eta/core/objects.py]: thanks for catching- this is fixed now

**Pull request merged**